### ISSUE TITLE
expectRevert fails on internal calls (jumps) if public calls are performed first

### ISSUE THREAD

**Issue by xenoliss** [foundry-rs/foundry#5454]
### Component

Forge

### Have you ensured that all of these are up to date?

- [x ] Foundry
- [ x] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (6e09b4e 2023-07-22T00:17:34.012434229Z)

### What command(s) is the bug in?

forge test --mc TestPoc

### Operating System

Linux

### Describe the bug

Hi,

Most of my application is composed of Solidity Libraries that exposed internal methods so that logic can be shared without inheritance accros different components.

While writing unit tests I noticed some weird behavior regarding the `expectRevert` cheatcode applied on internal methods. Here is a simple example PoC to demonstrate it: 
```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.21;

import "forge-std/Test.sol";

library MyLib {
  error MyError();

  function myMethodWithPublicCall(MyContract mc) internal returns (uint256 a) {
    a = mc.myFunc(12);
    revert MyError();
  }

  function myMethodWithoutPublicCall() internal returns (uint256 a) {
    a = 12;
    revert MyError();
  }
}

contract MyContract {
  function myFunc(uint256 a) external returns (uint256 _a) {
    _a = a;
  }
}

contract TestPoc is Test {
  MyContract private _mc;

  function setUp() public {
    _mc = new MyContract();
  }

  function testRevert_myMethodWithPublicCall_MyError() public {
    vm.expectRevert(MyLib.MyError.selector);
    MyLib.myMethodWithPublicCall(_mc);
  }

  function testRevert_myMethodWithoutPublicCall_MyError() public {
    vm.expectRevert(MyLib.MyError.selector);
    MyLib.myMethodWithoutPublicCall();
  }
}
```

Running `forge test --mc TestPoc` outputs the following:
```
Running 2 tests for test/TestPoc.t.sol:TestPoc
[FAIL. Reason: Call did not revert as expected] testRevert_myMethodWithPublicCall_MyError() (gas: 8390)
Traces:
  [8390] TestPoc::testRevert_myMethodWithPublicCall_MyError() 
    ├─ [0] VM::expectRevert(MyError()) 
    │   └─ ← ()
    ├─ [235] MyContract::myFunc(12) 
    │   └─ ← 12
    └─ ← "Call did not revert as expected"

[PASS] testRevert_myMethodWithoutPublicCall_MyError() (gas: 3164)
Test result: FAILED. 1 passed; 1 failed; 0 skipped; finished in 2.83ms
Ran 1 test suites: 1 tests passed, 1 failed, 0 skipped (2 total tests)

Failing tests:
Encountered 1 failing test in test/TestPoc.t.sol:TestPoc
[FAIL. Reason: Call did not revert as expected] testRevert_myMethodWithPublicCall_MyError() (gas: 8390)

Encountered a total of 1 failing tests, 1 tests succeeded
```

The test passes for the version that does not perform a public call **before** the error is raised. On the contrary if a public call is performed before (and this calls does not revert with the expected error), then the test fails.

What I would have imagined is `expectRevert` to behave the same way as `expectEmit`. `expectEmit` does not care if the very next event emitted it the one you are expecting. It only cares about two things:
1. The expected event is emitted **at some point** within the transaction
2. If several events are expected, you can omit some of them, but the ordering **must** match.

Would be really helpful to be able to use `expectRevert` so that if the expected error is raised at some point within the transaction, the test passes.

For now I'm using `testFail` but not a fan of it. I am aware I could maybe wrap all my libraries in contracts and exposes their internal methods but as I sad my entire application is 90% made of libs so that's quite painfull.

**Comment by mds1** [comment]
> I am aware I could maybe wrap all my libraries in contracts and exposes their internal methods but as I sad my entire application is 90% made of libs so that's quite painfull.

I know it's tedious to wrap like this, but unfortunately that is the recommended approach for now. Maybe we should have a codegen feature that can automatically generate these wrapper contracts? Let me know what you think, cc @gakonst @Evalir 

[This comment](https://github.com/foundry-rs/foundry/issues/5367#issuecomment-1635887313) provides references where you can read more about why we need to remove `expectRevert` support for jumps, I'd recommend giving that and the linked discussions a read.

The summary is: we can't reliably hook into jumps, and "supporting" expectReverts on jump was a bug that is now widely used, which is problematic because it has the following footgun:

```solidity
function testRevert_myMethodWithoutPublicCall_MyError() public {
  // Jump to library method.
  // This part works as expected and if the revert is seen the test passes.
  vm.expectRevert(MyLib.MyError.selector);
  MyLib.myMethodWithoutPublicCall();

  // Jump to library method again.
  // BUT execution never reaches here: it stops after the last expect revert was successful!
  vm.expectRevert(MyLib.MyError.selector);
  MyLib.myMethodWithoutPublicCall();
}
```

The reason you observe what you reported here is because in `testRevert_myMethodWithPublicCall_MyError` here's what happens:
1. `vm.expectRevert(MyLib.MyError.selector)`: We tell forge we expect a revert
2. `MyLib.myMethodWithPublicCall(_mc)`: Since `myMethodWithPublicCall` is an internal method, it's a JUMP and not a CALL
    1. Maybe some non-reverting internal logic is here, maybe not, doesn't matter much 
    1. `a = mc.myFunc(12)`: We make an external call and forge now expects the revert here! When this call succeeds, the scope of "where to expect a revert" is over and we never saw one, so the test fails. If the prior step (internal library logic) did revert, that's where we'd be relying on the accidental/footgun jumps support of `expectRevert`

Given the assumption that we can't expectReverts for jumps, we'd appreciate any feedback on how to simplify the test process. Manually wrapping large libraries is tedious, so maybe a codegen feature/convention is the best approach to facilitate this? Let me know what you think :)

**Comment by MathisGD** [comment]
Maybe with an elegant way to tell `testFail` the error message that it should expect it would be great. Something like: 

```solidity
function testFailMyLib() public {
    vm.failReason(MyLib.MyError.selector);
    MyLib.myMethodWithPublicCall();
}
```

What do you think ?

**Comment by xenoliss** [comment]
I ended up wrapping all my libs in my test files by doing the following in each test file:

1. Create an abstract `LibWrapper` contract at the top of the test file.
2. Expose `external` only methods that are wrapping my library internal methods (NOTE: some of the methods of the libs are taking storage references as params, for them I had to basically change the external interface to take the mapping key most of the time) with the following convention: `<method_name>_wrapper`.
3. Have my `TestContract` inherit my `LibWrapper`.
4. In my tests, replace all `Library.method()` calls to `this.method_wrapper()`.

All in all this is pretty repetitive but at the same time pretty easy to implement so maybe not that bad.

I have to say that I like @MathisGD idea a lot. But it kind of depend I guess if `testFail` was supposed to be deprecated or not in the future.

PS: Although having a `testFail` as described by @MathisGD might do the job for errors, wrapping libs would still be necessary when expecting events.
